File: examples/reference-implementations/typescript-client/src/discovery.ts
/**
 * OSSA Discovery Operations
 *
 * Client methods for discovering agents by taxonomy, capabilities, and compliance.
 */

import { OSSAClient } from './client.js';

export interface TaxonomyNode {
  domain: string;
  subdomains?: {
    [key: string]: {
      capabilities?: string[];
    };
  };
}

export interface TaxonomiesList {
  taxonomies: TaxonomyNode[];
}

export interface CapabilityDefinition {
  name: string;
  description: string;
  domain: string;
  subdomain?: string;
  example_agents: string[];
  schema?: {
    input: Record<string, unknown>;
    output: Record<string, unknown>;
  };
}

export interface CapabilitiesResponse {
  capabilities: CapabilityDefinition[];
}

export interface ComplianceProfile {
  id: string;
  name: string;
  description: string;
  requirements: Array<{
    category: string;
    requirement: string;
    validation: string;
  }>;
  agent_count: number;
}

export interface ComplianceProfilesResponse {
  profiles: ComplianceProfile[];
}

export interface DiscoveryFilters {
  domain?: string;
  subdomain?: string;
  capability?: string;
  compliance?: string[];
  min_rating?: number;
  verified_only?: boolean;
  has_dependencies?: boolean;
  license?: string;
}

export interface RecommendationRequest {
  use_case: string;
  requirements?: {
    compliance?: string[];
    budget?: string;
    performance?: 'low' | 'medium' | 'high';
    integration?: string[];
  };
  preferences?: {
    verified_only?: boolean;
    open_source_only?: boolean;
    min_rating?: number;
  };
}

export interface AgentRecommendation {
  agent: {
    publisher: string;
    name: string;
    version: string;
    description: string;
  };
  score: number;
  reasoning: string;
  compliance_match: string[];
  capability_match: string[];
  estimated_cost?: string;
}

export interface RecommendationResponse {
  recommendations: AgentRecommendation[];
  total_matches: number;
}

/**
 * Discovery operations client
 */
export class DiscoveryClient {
  constructor(private client: OSSAClient) {}

  /**
   * List available taxonomies
   */
  async listTaxonomies(): Promise<TaxonomiesList> {
    return this.client.request<TaxonomiesList>({
      method: 'GET',
      path: '/specification/taxonomies',
    });
  }

  /**
   * List capabilities by domain
   */
  async listCapabilities(domain?: string): Promise<CapabilitiesResponse> {
    return this.client.request<CapabilitiesResponse>({
      method: 'GET',
      path: '/specification/capabilities',
      query: domain ? { domain } : undefined,
    });
  }

  /**
   * Get capability definition
   */
  async getCapability(capabilityName: string): Promise<CapabilityDefinition> {
    return this.client.request<CapabilityDefinition>({
      method: 'GET',
      path: `/specification/capabilities/${capabilityName}`,
    });
  }

  /**
   * List compliance profiles
   */
  async listComplianceProfiles(): Promise<ComplianceProfilesResponse> {
    return this.client.request<ComplianceProfilesResponse>({
      method: 'GET',
      path: '/specification/compliance',
    });
  }

  /**
   * Get compliance profile details
   */
  async getComplianceProfile(profileId: string): Promise<ComplianceProfile> {
    return this.client.request<ComplianceProfile>({
      method: 'GET',
      path: `/specification/compliance/${profileId}`,
    });
  }

  /**
   * Discover agents by taxonomy
   */
  async discoverByTaxonomy(
    domain: string,
    subdomain?: string,
    capability?: string
  ): Promise<{
    taxonomy: { domain: string; subdomain?: string; capability?: string };
    agents: Array<{
      publisher: string;
      name: string;
      version: string;
      description: string;
      rating: number;
      downloads: number;
    }>;
    total: number;
  }> {
    const query: Record<string, string> = { domain };
    if (subdomain) query.subdomain = subdomain;
    if (capability) query.capability = capability;

    return this.client.request({
      method: 'GET',
      path: '/agents/discover/taxonomy',
      query,
    });
  }

  /**
   * Discover agents by compliance requirements
   */
  async discoverByCompliance(
    profiles: string[]
  ): Promise<{
    compliance_profiles: string[];
    agents: Array<{
      publisher: string;
      name: string;
      version: string;
      description: string;
      compliance_profiles: string[];
      verified: boolean;
    }>;
    total: number;
  }> {
    return this.client.request({
      method: 'GET',
      path: '/agents/discover/compliance',
      query: { profiles: profiles.join(',') },
    });
  }

  /**
   * Get agent recommendations based on use case
   */
  async getRecommendations(request: RecommendationRequest): Promise<RecommendationResponse> {
    return this.client.request<RecommendationResponse>({
      method: 'POST',
      path: '/agents/recommend',
      body: request,
    });
  }

  /**
   * Advanced discovery with multiple filters
   */
  async discover(filters: DiscoveryFilters): Promise<{
    filters: DiscoveryFilters;
    results: Array<{
      publisher: string;
      name: string;
      version: string;
      description: string;
      taxonomy: { domain: string; subdomain?: string; capability?: string };
      compliance_profiles: string[];
      rating: number;
      verified: boolean;
    }>;
    total: number;
  }> {
    return this.client.request({
      method: 'POST',
      path: '/agents/discover',
      body: filters,
    });
  }
}

File: examples/reference-implementations/python-client/ossa_client/discovery.py
"""
OSSA Discovery Operations

Client methods for discovering agents by taxonomy, capabilities, and compliance.
"""

from typing import Any, Dict, List, Optional
from .client import OSSAClient


class DiscoveryClient:
    """
    Discovery operations client for finding agents by various criteria.
    """

    def __init__(self, client: OSSAClient):
        self._client = client

    def list_taxonomies(self) -> Dict[str, Any]:
        """
        List available agent taxonomies.

        Returns:
            Available taxonomies with domains, subdomains, and capabilities
        """
        return self._client.request('GET', '/specification/taxonomies')

    def list_capabilities(self, domain: Optional[str] = None) -> Dict[str, Any]:
        """
        List available capabilities.

        Args:
            domain: Filter by domain

        Returns:
            Available capabilities with definitions and examples
        """
        query = {'domain': domain} if domain else None
        return self._client.request('GET', '/specification/capabilities', query=query)

    def get_capability(self, capability_name: str) -> Dict[str, Any]:
        """
        Get capability definition.

        Args:
            capability_name: Capability name

        Returns:
            Capability definition with schema and example agents
        """
        return self._client.request('GET', f'/specification/capabilities/{capability_name}')

    def list_compliance_profiles(self) -> Dict[str, Any]:
        """
        List available compliance profiles.

        Returns:
            Available compliance profiles (FedRAMP, HIPAA, SOC2, etc.)
        """
        return self._client.request('GET', '/specification/compliance')

    def get_compliance_profile(self, profile_id: str) -> Dict[str, Any]:
        """
        Get compliance profile details.

        Args:
            profile_id: Compliance profile ID

        Returns:
            Compliance profile requirements and agent count
        """
        return self._client.request('GET', f'/specification/compliance/{profile_id}')

    def discover_by_taxonomy(
        self,
        domain: str,
        subdomain: Optional[str] = None,
        capability: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Discover agents by taxonomy.

        Args:
            domain: Taxonomy domain
            subdomain: Taxonomy subdomain (optional)
            capability: Specific capability (optional)

        Returns:
            Agents matching the taxonomy criteria
        """
        query = {'domain': domain}
        if subdomain:
            query['subdomain'] = subdomain
        if capability:
            query['capability'] = capability

        return self._client.request('GET', '/agents/discover/taxonomy', query=query)

    def discover_by_compliance(self, profiles: List[str]) -> Dict[str, Any]:
        """
        Discover agents by compliance requirements.

        Args:
            profiles: List of compliance profile IDs

        Returns:
            Agents matching the compliance requirements
        """
        return self._client.request(
            'GET',
            '/agents/discover/compliance',
            query={'profiles': ','.join(profiles)},
        )

    def get_recommendations(
        self,
        use_case: str,
        requirements: Optional[Dict[str, Any]] = None,
        preferences: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """
        Get agent recommendations based on use case.

        Args:
            use_case: Description of the use case
            requirements: Requirements (compliance, budget, performance, integration)
            preferences: Preferences (verified_only, open_source_only, min_rating)

        Returns:
            Recommended agents with scores and reasoning
        """
        return self._client.request(
            'POST',
            '/agents/recommend',
            body={
                'use_case': use_case,
                'requirements': requirements,
                'preferences': preferences,
            },
        )

    def discover(self, filters: Dict[str, Any]) -> Dict[str, Any]:
        """
        Advanced discovery with multiple filters.

        Args:
            filters: Discovery filters (domain, subdomain, capability, compliance,
                    min_rating, verified_only, has_dependencies, license)

        Returns:
            Agents matching all filter criteria
        """
        return self._client.request('POST', '/agents/discover', body=filters)
